rclcpp_action/include/rclcpp_action/client_goal_handle.hpp

```diff
@@ -0,0 +1,64 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_ACTION__CLIENT_GOAL_HANDLE_HPP_
+#define RCLCPP_ACTION__CLIENT_GOAL_HANDLE_HPP_
+
+#include <rcl_action/action_client.h>
+
+#include <functional>
+#include <memory>
+
+#include "rclcpp_action/visibility_control.hpp"
+
+namespace rclcpp_action
+{
+// Forward declarations
+template<typename ACTION>
+class Client;
+
+template<typename ACTION>
+class ClientGoalHandle
+{
+public:
+  RCLCPP_ACTION_PUBLIC
+  virtual ~ClientGoalHandle();
+
+  /// TODO(sloretz) examples have this on client as `async_cancel_goal(goal_handle)`
+  std::future<bool>
+  async_cancel();
```

Comment:
@sloretz I do like the idea of calling cancel on the goal handle, but we have to support cancellation requests that expand multiple goals (e.g. cancel all goals). So I'd propose to keep this `async_cancel()` method here, plus `async_cancel_all_goals()` and `async_cancel_goals_before()` methods in the client, and maybe expose the low level method that just takes goal ID and timestamp for completeness.

Reply:
e79f5da8032e02e89d0a1b66ba379745f46a21fc